osu.Game.Rulesets.Catch/Skinning/LegacyComboCounter.cs

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using osu.Framework.Allocation;
+using osu.Framework.Bindables;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Game.Graphics.Sprites;
+using osu.Game.Graphics.UserInterface;
+using osu.Game.Rulesets.Catch.UI;
+using osu.Game.Screens.Play;
+using osu.Game.Skinning;
+using osuTK;
+using osuTK.Graphics;
+
+namespace osu.Game.Rulesets.Catch.Skinning
+{
+    internal class LegacyComboCounter : CompositeDrawable, ICatchComboCounter
+    {
+        private readonly ISkin skin;
+
+        private readonly string fontName;
+        private readonly float fontOverlap;
+
+        private readonly LegacyRollingCounter counter;
+        private LegacyRollingCounter lastExplosion;
+
+        public LegacyComboCounter(ISkin skin, string fontName, float fontOverlap)
+        {
+            this.skin = skin;
+
+            this.fontName = fontName;
+            this.fontOverlap = fontOverlap;
+
+            AutoSizeAxes = Axes.Both;
+
+            Alpha = 0f;
+            Anchor = Anchor.Centre;
+            Origin = Anchor.Centre;
+            Scale = new Vector2(0.8f);
+
+            InternalChild = counter = new LegacyRollingCounter(skin, fontName, fontOverlap)
+            {
+                Anchor = Anchor.Centre,
+                Origin = Anchor.Centre,
+            };
+        }
+
+        private IBindable<bool> isBreakTime;
+
+        [Resolved(canBeNull: true)]
+        private GameplayBeatmap beatmap { get; set; }
+
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+
+            isBreakTime = beatmap?.IsBreakTime.GetBoundCopy();
+            isBreakTime?.BindValueChanged(b =>
+            {
+                if (b.NewValue)
+                    this.FadeOut(400.0, Easing.OutQuint);
+            });
+        }
+
+        public void DisplayInitialCombo(int combo) => updateCombo(combo, null, true);
+        public void UpdateCombo(int combo, Color4? hitObjectColour) => updateCombo(combo, hitObjectColour, false);
+
+        private void updateCombo(int combo, Color4? hitObjectColour, bool immediate)
+        {
+            // Combo fell to zero, roll down and fade out the counter.
+            if (combo == 0)
+            {
+                counter.Current.Value = 0;
+                if (lastExplosion != null)
+                    lastExplosion.Current.Value = 0;
+
+                this.FadeOut(immediate ? 0.0 : 400.0, Easing.Out);
+                return;
+            }
```

Comment:
Sometimes this path doesn't seem to work so well. It seems that it is possible for the explosion swap to somehow interfere with this, which means that the first explosion after the miss that brought combo down to 0 the old combo count is displayed. See [reproduction video](https://drive.google.com/file/d/1cGS56djY7IiXW2gWCmiSxQX6iZEFTx8C/view?usp=sharing).

Also, shouldn't all this stuff be inside of a transform sequence of some sort?

Reply:
Hmm, good catch. Not quite sure what to do here, it sounds feasible to fix this, but it will make things pretty much complicated than now (the transforms to the new combo second pop up must be tracked, then somehow removed when the user miss). 

Thinking about just moving the `FinishTransforms(true)` above and not showing the explosion at all. (also side effect from this is that when missing after a beatmap break, the combo will display immediately and roll down, because `FadeOutFromOne` now must be used).

Reply:
I'm not sure how to respond to this comment unless I go and try to fix this myself. A proposal in code with some video would probably be better.